xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,398 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package gracefulswitch implements a graceful switch load balancer.
+package gracefulswitch
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/resolver"
+)
+
+var errBalancerClosed = errors.New("gracefulSwitchBalancer is closed")
+
+func newGracefulSwitchBalancer(cc balancer.ClientConn, opts balancer.BuildOptions) *gracefulSwitchBalancer {
+	return &gracefulSwitchBalancer{
+		cc:              cc,
+		bOpts:           opts,
+		scToSubBalancer: make(map[balancer.SubConn]balancer.Balancer),
+		closed:          grpcsync.NewEvent(),
+		pendingState: balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+		},
+	}
+}
+
+type gracefulSwitchBalancer struct {
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	mu               sync.Mutex
+	balancerCurrent  balancer.Balancer
+	balancerPending  balancer.Balancer
+	scToSubBalancer  map[balancer.SubConn]balancer.Balancer
+	pendingState     balancer.State
+	currentLbIsReady bool
+
+	// currentMu protects operations on the current balancer.
+	currentMu sync.Mutex
+
+	closed *grpcsync.Event
+}
+
+func (gsb *gracefulSwitchBalancer) updateState(bal balancer.Balancer, state balancer.State) {
+	if gsb.closed.HasFired() {
+		return
+	}
+
+	// I want this to not intersplice with other updateState() calls and for
+	// this to happen atomically (i.e. if this update isn't atomic writing to
+	// pending state concurrently and sending wrong one out for one
+	// updateState() call).
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	if bal == gsb.balancerPending {
+		// Cache the pending state and picker if you don't need to send it at
+		// this instant (i.e. the LB policy is connecting). This balancer can
+		// send it later on the current LB exiting READY.
+		gsb.pendingState = state
+		// "If the channel is currently in a state other than READY, the new
+		// policy will be swapped into place immediately." - Java
+		// implementation.
```

Comment:
It feels strange to quote the java implementation.  Re-state the reasons for things in a way that makes sense as a standalone comment here, please.